### ISSUE TITLE
Does components() in bokeh.embed work for widgets?

### ISSUE THREAD

**Issue by jaredsagendorf** [bokeh/bokeh#3893]
Unless I'm missing something, it seems that the components() function doesn't work for many widgets. Attached is a MWE that shows this for a few widgets. The only one which seems to function properly in that example is the DataTable. 

If this is the case, is there any other way to control the position of widgets without using VBoxForm() and Hbox() as in some of the app examples? I want to have a high-quality webpage with bokeh plots/widgets embedded in it, but in a layout of my choosing. Just smushing all the widgets and plots together into monolithic columns isn't going to cut it... I was hoping to wrap different plots/widgets in different \<div\>'s which I could then manipulate with some css layout.

I hope there's a good solution to this and working examples would be much appreciated. If this isn't possible then this is a very discouraging state of affairs...

```
from bokeh.models.widgets import Slider,CheckboxGroup,DataTable,MultiSelect,TableColumn
from bokeh.embed import components
from bokeh.util.browser import view
from bokeh.resources import INLINE
from bokeh.plotting import ColumnDataSource

from jinja2 import Template

# Some toy data
some_data = ColumnDataSource(
        data=dict(
            x=[0,1,2,3,4],
            y=['A','B','C','D','E'],
            name=['Aardvark','Bird','Cat','Dog','Emu']
        )
    )
columns = [
        TableColumn(field="name", title="Name"),
        TableColumn(field="x", title="X"),
        TableColumn(field="y", title="Y"),
    ]
# Create a widget
widget = Slider(start=0, end=10, value=1, step=.1, title="Stuff")
#widget = CheckboxGroup(labels=["One","Two","Three"],active=[1,1,1])
#widget = MultiSelect(title="Option:", value=["foo", "quux"],options=["foo", "bar", "baz", "quux"])
#widget = DataTable(source=some_data, columns=columns, width=400, height=280)

# Get JavaScript/HTML resources
script, div = components(widget)
js_resources = INLINE.render_js()
css_resources = INLINE.render_css()

template = Template('''<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>Widget</title>
        {{ js_resources }}
        {{ css_resources }}
    </head>
    <body>
        {{ div }}
        {{ script }}
    </body>
</html>
''')

filename = 'widget.html'

html = template.render(js_resources=js_resources,
        css_resources=css_resources,
        script=script,
        div=div)

with open(filename, 'w') as f:
    f.write(html)

view(filename)

```

**Comment by bryevdv** [comment]
Improving layout options is actually one of our highest immediate priorities, and there are people actively working on it currently right now. As Bokeh started as a plotting library primarily, I can say it really simply never occurred until recently that folks might want to have a page with just widgets as you've done here. So it's an edge case that was not tested for (because it was not even considered, really). OSS is an interactive process and we definitely rely on feedback from users to help find either bugs or new use-cases to support. 

All that said, an immediate workaround seems to be to just wrap the widget in a layout box of some sort first:
```
script, div = components(HBox(widget))
```
With that change your script works for me.

**Comment by canavandl** [comment]
It appears that the css selectors for the widgets are missing a "bk-" prefix, so the widget styles aren't being applied.

<img width="1118" alt="screen shot 2016-02-19 at 2 54 02 pm" src="https://cloud.githubusercontent.com/assets/2198981/13188703/ac972060-d718-11e5-9bae-b7ed669278d0.png">
<img width="764" alt="screen shot 2016-02-19 at 2 52 35 pm" src="https://cloud.githubusercontent.com/assets/2198981/13188704/aca413d8-d718-11e5-8790-32ed9da94c13.png">

That's definitely a bug.

**Comment by bryevdv** [comment]
An easy fix hopefully!

**Comment by canavandl** [comment]
ref PR #3797

**Comment by jaredsagendorf** [comment]
Thanks for the quick replies. The HBox wrapper is a great tip, but there still seems to be a few kinks - the text in MultiSelect seems to be missing, and the CheckboxGroup doesn't choose the correct initial selections defined by active=[1,1,1] (at least as of 0.11.0). 

I hope the layout features are added in the next release. I think the ability to input a template HTML file, and get back the plots/widgets where you want within a web page would be a key feature to make Bokeh competitive with other visualization libraries out there, especially those (few) geared towards python.

**Comment by canavandl** [comment]
And for some reason, wrapping the widgets in a HBox (or similar) adds the prefix, so that the css selectors match?

<img width="1114" alt="screen shot 2016-02-19 at 3 29 49 pm" src="https://cloud.githubusercontent.com/assets/2198981/13189538/af5b209e-d71d-11e5-8359-2979f70cbe82.png">

**Comment by canavandl** [comment]
It appears that wrapping child roots (like a widget) in a component causes the "bk-" prefix to be added during the build_views call:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/common/build_views.coffee#L48

**Comment by mistakevin** [comment]
This was the subject of this PR https://github.com/bokeh/bokeh/pull/2666

**Comment by canavandl** [comment]
Thanks @mistakevin

**Comment by birdsarah** [comment]
Confirm in #4392.

**Comment by bryevdv** [comment]
@birdsarah FYI the above code "works" for every case except slider, which does not completely render. However, if I put the slider in a `Row` or a `WidgetBox` it works as expected.

**Comment by bryevdv** [comment]
more info, with this code:
```
# Create a widget
w1 = Slider(start=0, end=10, value=1, step=.1, title="Stuff")
w2 = CheckboxGroup(labels=["One","Two","Three"],active=[1,1,1])
w3 = MultiSelect(title="Option:", value=["foo", "quux"],options=["foo", "bar", "baz", "quux"])
w4 = DataTable(source=some_data, columns=columns, width=400, height=280)

# Get JavaScript/HTML resources
script, div = components(WidgetBox(w1, w2, w3, w4, width=500))
```
I see

<img width="581" alt="screen shot 2016-05-31 at 5 14 58 pm" src="https://cloud.githubusercontent.com/assets/1078448/15692307/48130448-2753-11e6-9a99-c8072e72fc7e.png">

which is nice tho perhaps the "embed just a widget" case should also work, even if that means auto-wrapping in some layout.

**Comment by bryevdv** [comment]
@jaredsagendorf FYI, small issue with slider aside, here is an updated script that renders three widgets directly into three separate divs (on current GH master):
```
from bokeh.models.widgets import Slider,CheckboxGroup,DataTable,MultiSelect,TableColumn
from bokeh.models.layouts import WidgetBox
from bokeh.embed import components
from bokeh.util.browser import view
from bokeh.resources import INLINE
from bokeh.plotting import ColumnDataSource, curdoc

from jinja2 import Template

# Some toy data
some_data = ColumnDataSource(
        data=dict(
            x=[0,1,2,3,4],
            y=['A','B','C','D','E'],
            name=['Aardvark','Bird','Cat','Dog','Emu']
        )
    )
columns = [
        TableColumn(field="name", title="Name"),
        TableColumn(field="x", title="X"),
        TableColumn(field="y", title="Y"),
    ]
# Create a widget
w1 = Slider(start=0, end=10, value=1, step=.1, title="Stuff")
w2 = CheckboxGroup(labels=["One","Two","Three"],active=[1,1,1])
w3 = MultiSelect(title="Option:", value=["foo", "quux"],options=["foo", "bar", "baz", "quux"])
w4 = DataTable(source=some_data, columns=columns, height=280)

curdoc().add_root(w2)
curdoc().add_root(w3)
curdoc().add_root(w4)

# Get JavaScript/HTML resources
script, (div1, div2, div3) = components((w2, w3, w4))
js_resources = INLINE.render_js()
css_resources = INLINE.render_css()

template = Template('''<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>Widget</title>
        {{ js_resources }}
        {{ css_resources }}
    </head>
    <body>
        {{ div1 }}
        {{ div2 }}
        {{ div3 }}
        {{ script }}
    </body>
</html>
''')

filename = 'widget.html'

html = template.render(js_resources=js_resources,
        css_resources=css_resources,
        script=script,
        div1=div1, div2=div2, div3=div3)

with open(filename, 'w') as f:
    f.write(html)
```
I realize this issue was opened some time ago, but the layout overhaul was a very extensive undertaking. Hopefully with the next `0.12` release all of your needs can be met.

**Comment by birdsarah** [comment]
Marking this as done. Fixed in #4343. Have opened a separate issue for slider rendering: #4507
